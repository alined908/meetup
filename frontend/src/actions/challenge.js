import * as types from "../constants/action-types";
import { axiosClient } from "../accounts/axiosClient";

export const getRestaurants = () => async dispatch => {
    dispatch({type: types.GET_RESTAURANTS_REQUEST})
    try {
        const response = await axiosClient.get(`/api/restaurants/challenge/`)
        dispatch({type: types.GET_RESTAURANTS_SUCCESS, payload: response.data})
    } catch(e) {
        dispatch({type: types.GET_RESTAURANTS_ERROR})
    }
}

export const getCategories = () => async dispatch => {
    dispatch({type: types.GET_CATEGORIES_REQUEST})
    try {
        const response = await axiosClient.get(`/api/categories/challenge/`)
        dispatch({type: types.GET_CATEGORIES_SUCCESS, payload: response.data})
    } catch(e) {
        dispatch({type: types.GET_CATEGORIES_ERROR})
    }
}

export const constructMatrix = (restaurants, numCategories) => {
    let matrix = {}
    restaurants.map((entity) => {
        matrix[entity.id] = Array(numCategories).fill(false)
    })
    console.log(matrix)
    return {
        type: types.CONSTRUCT_MATRIX,
        payload: matrix
    }
}

export const reconstructMatrix = newMatrix => {
    return {
        type: types.RECONSTRUCT_MATRIX,
        payload: newMatrix
    }
}

export const undoMatrix = () => {
    return {
        type: types.UNDO_MATRIX
    }
}

export const redoMatrix = () => {
    return {
        type: types.REDO_MATRIX
    }
}