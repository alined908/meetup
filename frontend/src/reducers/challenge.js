import * as types from "../constants/action-types";

export const defaultState = {
    restaurants: [],
    categories: [],
    matrix: {},
    matrixHistory: [],
    page: 0,
    isMatrixConstructed: false
}

export default function challengeReducer(state=defaultState, action){

    switch(action.type){
        case types.GET_RESTAURANTS_REQUEST:
            return {...state, isRestaurantsFetching: true}
        case types.GET_RESTAURANTS_SUCCESS:
            return {...state, restaurants: action.payload, isRestaurantsFetching: false, isRestaurantsInitialized: true}
        case types.GET_RESTAURANTS_ERROR:
            return {...state, isRestaurantsFetching: false, isRestaurantsInitialized: false}
        case types.GET_CATEGORIES_REQUEST:
            return {...state, isCategoriesFetching: true}
        case types.GET_CATEGORIES_SUCCESS:
            return {...state, categories: action.payload, isCategoriesFetching: false, isCategoriesInitialized: true}
        case types.GET_CATEGORIES_ERROR:
            return {...state, isCategoriesFetching: false, isCategoriesInitialized: false}
        case types.CONSTRUCT_MATRIX:
            return {...state, matrix: action.payload, isMatrixConstructed: true, matrixHistory: [action.payload]}
        case types.RECONSTRUCT_MATRIX:
            return {...state, matrix: action.payload, matrixHistory: [...state.matrixHistory.splice(0, state.page + 1), action.payload], page: state.page + 1}
        case types.UNDO_MATRIX:
            return {...state, page: state.page - 1, matrix: state.matrixHistory[state.page - 1]}
        case types.REDO_MATRIX:
            return {...state, page: state.page + 1, matrix: state.matrixHistory[state.page + 1]} 
        default:
            return state
    }
}