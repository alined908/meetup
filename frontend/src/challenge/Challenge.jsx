import React, {Component} from 'react'
import {axiosClient} from '../accounts/axiosClient'
import styles from "../styles/challenge.module.css"
import {DragDropContext, Droppable, Draggable} from 'react-beautiful-dnd'
import {CircularProgress, Button, IconButton} from '@material-ui/core'
import CloseIcon from '@material-ui/icons/Close';
import {connect} from 'react-redux'
import {isEqual} from 'lodash'
import {addGlobalMessage} from '../actions'
import {Undo as UndoIcon, Redo as RedoIcon} from '@material-ui/icons'
import {getRestaurants, getCategories, constructMatrix, reconstructMatrix, undoMatrix, redoMatrix} from '../actions/challenge.js'

const getStyle = (style, snapshot) => {
    if (!snapshot.isDragging) return {};
    if (!snapshot.isDropAnimating) {
      return style;
    }
    return {
      ...style,
      transitionDuration: `0.001s`,
    };
  }

class Card extends Component {
    constructor(props){
        super(props);
        this.state = {
            isHovered: false
        }
    }

    render () {
        const restaurant = this.props.restaurant
        return (
            <div className={`${styles.rstCard} ${this.state.isHovered ? styles.closeHovered : ""}`}>
                {!this.props.isMenu && 
                    <CloseIcon
                        size="small" 
                        color="secondary"
                        className={styles.closeIcon}
                        onClick={this.props.handleDelete} 
                        onMouseEnter={() => this.setState({isHovered: true})} 
                        onMouseLeave={() => this.setState({isHovered: false})}
                    />
                }
                <div className={styles.rstCardTop}>
                    <img className={styles.rstImg} src={restaurant.yelp_image} />
                </div>
                <div className={styles.rstCardBottom}>
                    <span>{restaurant.name}</span>
                    <div className={styles.rstCardCategories}>
                        {this.props.isMenu && this.props.categories.map((category, index) => 
                            this.props.filled[index] && <img
                                key={category.id} 
                                className={styles.categoryPicture} 
                                src={`${process.env.REACT_APP_S3_STATIC_URL}/static/category/${category.api_label}.png`}
                            />
                        )}
                    </div>
                </div>
            </div>
        )
    }
}

class Restaurant extends Component {

    render () {
        const restaurant = this.props.restaurant

        return (
            <Draggable draggableId={this.props.isMenu ? `restaurant-${restaurant.id}`: `category-${this.props.index}`} index={this.props.index}>
                {(provided, snapshot) => (
                    <React.Fragment>
                        <div 
                            ref={provided.innerRef} 
                            {...provided.draggableProps}
                            {...provided.dragHandleProps}
                            className={styles.restaurant}
                            style={getStyle(provided.draggableProps.style, snapshot)}
                        >   
                            <Card 
                                restaurant={restaurant} 
                                isMenu={this.props.isMenu}
                                handleDelete={this.props.handleDelete}
                                filled={this.props.filled}
                                categories={this.props.categories}
                            />
                        </div>
                        {snapshot.isDragging && this.props.isMenu && (
                            <Card 
                                restaurant={restaurant} 
                                isMenu={this.props.isMenu}
                                handleDelete={this.props.handleDelete}
                                filled={this.props.filled}
                                categories={this.props.categories}
                            />
                        )}
                    </React.Fragment>
                )}

            </Draggable>
        )
    }
}


const copy = (filled, source, destination, numCategories) => {

    const restaurantID = Number(source.droppableId.split("-")[1])
    const column = Number(destination.droppableId.split("-")[1]) % numCategories

    let oldRow = [...filled[restaurantID]]
    oldRow[column] = true

    return {...filled, [restaurantID]: [...oldRow]}
}

const move = (filled, source, destination, numCategories) => {
    const sourceCategory = Number(source.droppableId.split("-")[1])
    const restaurantID = Math.floor(sourceCategory / numCategories)
    const oldColumn = sourceCategory % numCategories
    const newColumn = Number(destination.droppableId.split("-")[1]) % numCategories

    let oldRow = [...filled[restaurantID]]
    oldRow[oldColumn] = false
    oldRow[newColumn] = true

    return {...filled, [restaurantID]: [...oldRow]}
}   

class RestaurantGrid extends Component {

    handleDelete = (restaurant, column) => {
        let oldRow = [...this.props.matrix[restaurant]]
        oldRow[column] = false
        this.props.reconstructMatrix({ ...this.props.matrix, [restaurant]: [...oldRow]})
        console.log({ ...this.props.matrix, [restaurant]: [...oldRow]})
    }

    onDragEnd = result => {
        const {source, destination} = result;
        if (!destination){
            return;
        }  

        const numCategories = this.props.categories.length;
        var filled;

        if (source.droppableId.includes("restaurant")) {
            filled = copy(this.props.matrix, source, destination, numCategories)
        } else if (source.droppableId.includes("category")){
            filled = move(this.props.matrix, source, destination, numCategories)
        }

        if (!isEqual(this.props.matrix, filled)) {
            this.props.reconstructMatrix(filled)
        }
        
    }

    render () {
        const restaurants = this.props.restaurants
        const categories = this.props.categories

        return (
            <DragDropContext onDragEnd={this.onDragEnd}>
                {restaurants.map((restaurant, index) => 
                    <div className={styles.restaurantRow} key={restaurant.id}>
                        <Droppable droppableId={`restaurant-${restaurant.id}`} isDropDisabled={true}>
                            {(provided, snapshot) => (
                                <div ref={provided.innerRef} {...provided.droppableProps} className={styles.restaurantWrapper}>
                                    <Restaurant 
                                        isMenu={true}
                                        filled={this.props.matrix[restaurant.id]}
                                        restaurant={restaurant}
                                        categories={categories}
                                        index={index}
                                    />
                                    <span style={{display: "none"}}>
                                        {provided.placeholder}
                                    </span>
                                </div>
                            )}
                        </Droppable>
                        <div className={styles.columns}>
                            {categories.map((category, index) => 
                                <Droppable key={category.id} droppableId={`category-${restaurant.id * categories.length + index}`}>
                                    {(provided, snapshot) => (
                                        <div ref={provided.innerRef} className={styles.block}>
                                            {this.props.matrix[restaurant.id][index] ?  
                                                <Restaurant 
                                                    isMenu={false}
                                                    restaurant={restaurant}
                                                    handleDelete={() => this.handleDelete(restaurant.id, index)}
                                                    index={restaurant.id * categories.length + index} />
                                                :   
                                                <div className={styles.category}/>
                                            }
                                            <span style={{display: "none"}}>
                                                {provided.placeholder}
                                            </span>
                                        </div>  
                                    )}
                                </Droppable>
                            )}
                        </div>
                    </div>
                )}
            </DragDropContext>
        )
    }
}

class Challenge extends Component {


    async componentDidMount() {
        await Promise.all([this.props.getRestaurants(), this.props.getCategories()])
        this.props.constructMatrix(this.props.restaurants, this.props.categories.length)
    }

    saveMatrix = () => {
        localStorage.setItem('matrix', JSON.stringify(this.props.matrix))
        this.props.addGlobalMessage("success", "Saved to LocalStorage.")
    }

    loadMatrix = () => {
        let matrix = localStorage.getItem('matrix')

        if (matrix === null){
            this.props.addGlobalMessage("error", "No matrix in LocalStorage")
        } else{
            this.props.reconstructMatrix(JSON.parse(matrix))
        }
        
    }

    render () {

        return (
            <>
                {!this.props.isFetching && this.props.isInitialized ?
                    <div className={`${styles.outerWrapper}`}>
                        <div className={styles.topWrapper}>
                            <div className={styles.restaurantsWrapper}>
                                <div className={styles.header}>
                                    Restaurants
                                </div>
                                <div className={styles.restaurantIndent}/>
                            </div>
                            <div className={styles.categoriesWrapper}>
                                <div className={styles.header}>
                                    <div>
                                        <IconButton color="primary" onClick={this.props.undoMatrix} disabled={!this.props.canUndo}>
                                            <UndoIcon/>
                                        </IconButton>
                                        <IconButton color="primary" onClick={this.props.redoMatrix} disabled={!this.props.canRedo}>
                                            <RedoIcon/>
                                        </IconButton>
                                    </div>
                                    <div>
                                        Categories
                                    </div>
                                    <div>
                                        <Button variant='contained' color="primary" onClick={this.saveMatrix} className={styles.save}> 
                                            Save
                                        </Button>
                                        <Button variant='contained' color="primary" onClick={this.loadMatrix}> 
                                            Load
                                        </Button>
                                    </div>
                                </div>
                                <div className={styles.categories}>
                                    {this.props.categories.map((category, index) => 
                                        <div key={index} className={styles.categoryHeaderWrapper}>
                                            <div className={styles.categoryHeader}>
                                                <img 
                                                    className={styles.categoryPicture} 
                                                    src={`${process.env.REACT_APP_S3_STATIC_URL}/static/category/${category.api_label}.png`}
                                                />
                                                {category.label}
                                            </div>
                                        </div>
                                    )}
                                </div>
                            </div>
                        </div>
                        <div className={styles.bottomWrapper}>
                            <div className={styles.restaurants}>
                                <RestaurantGrid 
                                    restaurants={this.props.restaurants} 
                                    categories={this.props.categories}
                                    constructMatrix={this.props.constructMatrix}
                                    reconstructMatrix={this.props.reconstructMatrix}
                                    matrix={this.props.matrix}
                                    matrixHistory={this.props.matrixHistory}
                                />
                            </div>
                        </div>
                    </div>
                    :
                    <div className="loading">
                        <CircularProgress size={30}/>
                    </div>

                }       
            </>
        )
    }
}

const mapStateToProps = state => {
    return {
        restaurants: state.challenge.restaurants,
        categories: state.challenge.categories,
        isFetching: state.challenge.isRestaurantsFetching || state.challenge.isCategoriesFetching,
        isInitialized: state.challenge.isRestaurantsInitialized && state.challenge.isCategoriesInitialized && state.challenge.isMatrixConstructed,
        matrix: state.challenge.matrix,
        page: state.challenge.page,
        matrixHistory: state.challenge.matrixHistory,
        canUndo: state.challenge.page > 0,
        canRedo: state.challenge.page < state.challenge.matrixHistory.length - 1,
    }
}

const mapDispatchToProps = {
    getRestaurants,
    getCategories,
    constructMatrix,
    reconstructMatrix,
    undoMatrix,
    redoMatrix,
    addGlobalMessage
}

export default connect(mapStateToProps, mapDispatchToProps)(Challenge) 