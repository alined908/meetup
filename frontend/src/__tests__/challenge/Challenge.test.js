import React from "react";
import { shallow, mount } from "enzyme";
import { RestaurantGrid, Restaurant, Card, Challenge } from "../../components/challenge/Challenge"

//Mocks
const matrix = {
    1: [false, false, false, false],
    2: [false, false, false, false],
    3: [false, false, false, false],
    4: [false, false, false, false],
}

const categories = [
    {id: 1, api_label: "dessert", label: "Dessert"},
    {id: 2, api_label: "pizza", label: "Pizza"},
    {id: 3, api_label: "chinese", label: "Chinese"},
    {id: 4, api_label: "korean", label: "Korean"}
]

const restaurants = [
    {id: 1, name: "Pizza place"},
    {id: 2, name: "Dessert place"},
    {id: 3, name: "Chinese place"},
    {id: 4, name: "Korean place"}
]

//Components
const cardProps = {
    filled: matrix[1],
    categories,
    restaurant: restaurants[0],
    handleDelete: jest.fn()
}

describe("Card", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing given props", () => {
        const wrapper = shallow(<Card {...cardProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('does not render closeIcon if part of menu', () => {
        const updatedProps = {...cardProps, isMenu: true}
        const wrapper = shallow(<Card {...updatedProps} />);
        const closeIcon = wrapper.find('.closeIcon')
        expect(closeIcon).toHaveLength(0)
    })

    it('renders closeIcon if not part of menu and handles onClick', () => {
        const updatedProps = {...cardProps, isMenu: false}
        const wrapper = shallow(<Card {...updatedProps} />);
        const closeIcon = wrapper.find('.closeIcon')
        expect(closeIcon).toHaveLength(1)
        closeIcon.simulate('click')
        expect(updatedProps.handleDelete.mock.calls.length).toBe(1);
    })

    it ('renders categories if part of menu', () => {
        let newMatrix = [...matrix[1]]
        newMatrix[0] = true
        const updatedProps = {...cardProps, isMenu: true, filled: newMatrix}
        const wrapper = shallow(<Card {...updatedProps} />);
        const categories = wrapper.find('.categoryPicture')
        expect(categories).toHaveLength(1)
    })
})

const restaurantProps = {
    restaurant: restaurants[0],
    isMenu: false,
    filled: matrix[1],
    categories,
    handleDelete: jest.fn(),
    index: 0,
}

describe("Restaurant", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing given props", () => {
        const wrapper = shallow(<Restaurant {...restaurantProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('displays draggableId as category-X if not part of menu', () => {
        const wrapper = shallow(<Restaurant {...restaurantProps} />);
        const draggable = wrapper.find('PublicDraggable')
        expect(draggable).toHaveLength(1)
        expect(draggable.prop('draggableId')).toEqual('category-0')
    })

    it('displays draggableId as restaurant-X if part of menu', () => {
        const updatedProps = {...restaurantProps, isMenu: true}
        const wrapper = shallow(<Restaurant {...updatedProps} />);
        const draggable = wrapper.find('PublicDraggable')
        expect(draggable).toHaveLength(1)
        expect(draggable.prop('draggableId')).toEqual('restaurant-1')
    })
})

const gridProps = {
    matrix,
    categories,
    restaurants,
    reconstructMatrix: jest.fn()
}

describe("RestaurantGrid", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing given props", () => {
        const wrapper = shallow(<RestaurantGrid {...gridProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('constructs numCategories + 1 * numRestaurants droppable zones', () => {
        const wrapper = shallow(<RestaurantGrid {...gridProps} />);
        const droppables = wrapper.find('Connect(Droppable)')
        expect(droppables).toHaveLength((gridProps.categories.length + 1) * gridProps.restaurants.length)
    })

    it('makes call to reconstructMatrix if handleDelete called', () => {
        const wrapper = shallow(<RestaurantGrid {...gridProps} />);
        wrapper.instance().handleDelete(1, 0)
        expect(gridProps.reconstructMatrix.mock.calls.length).toBe(1)
    })

    it('reconstructs matrix given restaurant source and category destination', () => {
        const wrapper = shallow(<RestaurantGrid {...gridProps} />);
        const source = {droppableId: 'restaurant-1'}
        const destination = {droppableId: 'category-1'}
        const result = {source, destination}
        wrapper.instance().onDragEnd(result)
        const params = gridProps.reconstructMatrix.mock.calls[0][0]
        const expectedResult = {
            ...matrix,
            [1]: [false, true, false , false]
        }
        expect(gridProps.reconstructMatrix.mock.calls.length).toBe(1)
        expect(params).toEqual(expectedResult)
    })

    it('reconstructs matrix given category source and category destination', () => {
        const wrapper = shallow(<RestaurantGrid {...gridProps} />);
        const source = {droppableId: 'category-4'}
        const destination = {droppableId: 'category-5'}
        const result = {source, destination}
        wrapper.instance().onDragEnd(result)
        const params = gridProps.reconstructMatrix.mock.calls[0][0]
        const expectedResult = {
            ...matrix,
            [1]: [false, true, false , false]
        }
        expect(gridProps.reconstructMatrix.mock.calls.length).toBe(1)
        expect(params).toEqual(expectedResult)
    })
})

const challengeProps = {
    restaurants,
    categories,
    matrix,
    isFetching: false,
    isInitialized: true,
    getRestaurants: jest.fn(),
    getCategories: jest.fn(),
    addGlobalMessage: jest.fn(),
    constructMatrix: jest.fn(),
    reconstructMatrix: jest.fn(),
    undoMatrix: jest.fn(),
    redoMatrix: jest.fn()
}

describe("Challenge", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders without crashing given props", () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ('should show loading state if not initialized or fetching', () => {
        const updatedProps = {...challengeProps, isFetching: true}
        const wrapper = shallow(<Challenge {...updatedProps} />);
        const loading = wrapper.find('.loading')
        expect(loading).toHaveLength(1)
    })

    it('makes calls to getRestaurants and getCategories on component mount', () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        expect(challengeProps.getRestaurants.mock.calls.length).toBe(1)
        expect(challengeProps.getCategories.mock.calls.length).toBe(1)
    })

    it('makes call to saveMatrix if save Button is clicked', () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        const save = wrapper.find('.save')
        expect(save).toHaveLength(1)
        save.simulate('click')
        expect(challengeProps.addGlobalMessage.mock.calls.length).toBe(1)
    })

    it('makes call to loadMatrix if load Button is clicked', () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        const load = wrapper.find('.load')
        expect(load).toHaveLength(1)
        load.simulate('click')
        expect(challengeProps.reconstructMatrix.mock.calls.length).toBe(1)
    })

    it ('makes call to undoMatrix if undo Button is clicked', () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        const undo = wrapper.find('[data-testid="undo"]')
        expect(undo).toHaveLength(1)
        undo.simulate('click')
        expect(challengeProps.undoMatrix.mock.calls.length).toBe(1)
    })

    it ('makes call to redoMatrix if redo Button is clicked', () => {
        const wrapper = shallow(<Challenge {...challengeProps} />);
        console.log(wrapper.debug())
        const redo = wrapper.find('[data-testid="redo"]')
        expect(redo).toHaveLength(1)
        redo.simulate('click')
        expect(challengeProps.redoMatrix.mock.calls.length).toBe(1)
    })
})