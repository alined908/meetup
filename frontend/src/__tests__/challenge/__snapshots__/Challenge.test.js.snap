// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders without crashing given props 1`] = `
<div
  className="rstCard "
>
  <CloseIcon
    className="closeIcon"
    color="secondary"
    onClick={[MockFunction]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size="small"
  />
  <div
    className="rstCardTop"
  >
    <img
      className="rstImg"
    />
  </div>
  <div
    className="rstCardBottom"
  >
    <span>
      Pizza place
    </span>
    <div
      className="rstCardCategories"
    />
  </div>
</div>
`;

exports[`Challenge renders without crashing given props 1`] = `
<Fragment>
  <div
    className="outerWrapper"
  >
    <div
      className="topWrapper"
    >
      <div
        className="restaurantsWrapper"
      >
        <div
          className="header"
        >
          Restaurants
        </div>
        <div
          className="restaurantIndent"
        />
      </div>
      <div
        className="categoriesWrapper"
      >
        <div
          className="header"
        >
          <div>
            <WithStyles(ForwardRef(IconButton))
              color="primary"
              data-testid="undo"
              disabled={true}
              onClick={[MockFunction]}
            >
              <UndoIcon />
            </WithStyles(ForwardRef(IconButton))>
            <WithStyles(ForwardRef(IconButton))
              color="primary"
              data-testid="redo"
              disabled={true}
              onClick={[MockFunction]}
            >
              <RedoIcon />
            </WithStyles(ForwardRef(IconButton))>
          </div>
          <div>
            Categories
          </div>
          <div>
            <WithStyles(ForwardRef(Button))
              className="save"
              color="primary"
              onClick={[Function]}
              variant="contained"
            >
              Save
            </WithStyles(ForwardRef(Button))>
            <WithStyles(ForwardRef(Button))
              className="load"
              color="primary"
              onClick={[Function]}
              variant="contained"
            >
              Load
            </WithStyles(ForwardRef(Button))>
          </div>
        </div>
        <div
          className="categories"
        >
          <div
            className="categoryHeaderWrapper"
            key="0"
          >
            <div
              className="categoryHeader"
            >
              <img
                className="categoryPicture"
                src="undefined/static/category/dessert.png"
              />
              Dessert
            </div>
          </div>
          <div
            className="categoryHeaderWrapper"
            key="1"
          >
            <div
              className="categoryHeader"
            >
              <img
                className="categoryPicture"
                src="undefined/static/category/pizza.png"
              />
              Pizza
            </div>
          </div>
          <div
            className="categoryHeaderWrapper"
            key="2"
          >
            <div
              className="categoryHeader"
            >
              <img
                className="categoryPicture"
                src="undefined/static/category/chinese.png"
              />
              Chinese
            </div>
          </div>
          <div
            className="categoryHeaderWrapper"
            key="3"
          >
            <div
              className="categoryHeader"
            >
              <img
                className="categoryPicture"
                src="undefined/static/category/korean.png"
              />
              Korean
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bottomWrapper"
    >
      <div
        className="restaurants"
      >
        <RestaurantGrid
          categories={
            Array [
              Object {
                "api_label": "dessert",
                "id": 1,
                "label": "Dessert",
              },
              Object {
                "api_label": "pizza",
                "id": 2,
                "label": "Pizza",
              },
              Object {
                "api_label": "chinese",
                "id": 3,
                "label": "Chinese",
              },
              Object {
                "api_label": "korean",
                "id": 4,
                "label": "Korean",
              },
            ]
          }
          matrix={
            Object {
              "1": Array [
                false,
                false,
                false,
                false,
              ],
              "2": Array [
                false,
                false,
                false,
                false,
              ],
              "3": Array [
                false,
                false,
                false,
                false,
              ],
              "4": Array [
                false,
                false,
                false,
                false,
              ],
            }
          }
          reconstructMatrix={[MockFunction]}
          restaurants={
            Array [
              Object {
                "id": 1,
                "name": "Pizza place",
              },
              Object {
                "id": 2,
                "name": "Dessert place",
              },
              Object {
                "id": 3,
                "name": "Chinese place",
              },
              Object {
                "id": 4,
                "name": "Korean place",
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`Restaurant renders without crashing given props 1`] = `
<PublicDraggable
  draggableId="category-0"
  index={0}
>
  <Component />
</PublicDraggable>
`;

exports[`RestaurantGrid renders without crashing given props 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <div
    className="restaurantRow"
    key="1"
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="restaurant-1"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={true}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
    <div
      className="columns"
    >
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-4"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="1"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-5"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="2"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-6"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="3"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-7"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="4"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
    </div>
  </div>
  <div
    className="restaurantRow"
    key="2"
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="restaurant-2"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={true}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
    <div
      className="columns"
    >
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-8"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="1"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-9"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="2"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-10"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="3"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-11"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="4"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
    </div>
  </div>
  <div
    className="restaurantRow"
    key="3"
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="restaurant-3"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={true}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
    <div
      className="columns"
    >
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-12"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="1"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-13"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="2"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-14"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="3"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-15"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="4"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
    </div>
  </div>
  <div
    className="restaurantRow"
    key="4"
  >
    <Connect(Droppable)
      direction="vertical"
      droppableId="restaurant-4"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={true}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
    <div
      className="columns"
    >
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-16"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="1"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-17"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="2"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-18"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="3"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
      <Connect(Droppable)
        direction="vertical"
        droppableId="category-19"
        getContainerForClone={[Function]}
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        key="4"
        mode="standard"
        renderClone={null}
        type="DEFAULT"
      >
        <Component />
      </Connect(Droppable)>
    </div>
  </div>
</DragDropContext>
`;
